File: difftest/port/canon.go
package main

// Twin of ../pi/canon.mjs. Keep the two in lockstep: any change to what is
// normalized must land on both sides, or the differential starts comparing
// apples to oranges.

import (
	"bytes"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)

// rawNumber is a JSON number kept as its original literal text, so 1024 vs
// 1024.0 vs 1e3 stays visible instead of being laundered through a float.
type rawNumber string

// parsePreservingNumbers decodes JSON into any, with numbers as rawNumber and
// objects as orderedObject (insertion order preserved).
func parsePreservingNumbers(data []byte) (any, error) {
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	v, err := parseValue(dec)
	if err != nil {
		return nil, err
	}
	return v, nil
}

// orderedObject is a JSON object that remembers its key order.
type orderedObject struct {
	keys []string
	vals map[string]any
}

func parseValue(dec *json.Decoder) (any, error) {
	tok, err := dec.Token()
	if err != nil {
		return nil, err
	}
	return parseFromToken(dec, tok)
}

func parseFromToken(dec *json.Decoder, tok json.Token) (any, error) {
	switch t := tok.(type) {
	case json.Delim:
		switch t {
		case '{':
			obj := &orderedObject{vals: map[string]any{}}
			for dec.More() {
				kt, err := dec.Token()
				if err != nil {
					return nil, err
				}
				key, ok := kt.(string)
				if !ok {
					return nil, fmt.Errorf("object key is %T, want string", kt)
				}
				val, err := parseValue(dec)
				if err != nil {
					return nil, err
				}
				obj.keys = append(obj.keys, key)
				obj.vals[key] = val
			}
			if _, err := dec.Token(); err != nil { // closing }
				return nil, err
			}
			return obj, nil
		case '[':
			arr := []any{}
			for dec.More() {
				val, err := parseValue(dec)
				if err != nil {
					return nil, err
				}
				arr = append(arr, val)
			}
			if _, err := dec.Token(); err != nil { // closing ]
				return nil, err
			}
			return arr, nil
		}
		return nil, fmt.Errorf("unexpected delimiter %v", t)
	case json.Number:
		return rawNumber(t.String()), nil
	default:
		return tok, nil // string, bool, nil
	}
}

// encodeString matches JSON.stringify (ES2019 well-formed): escape the
// mandatory control set and lone surrogates, nothing else. In particular
// U+2028/U+2029 are NOT escaped, and neither are <, >, & — Go's encoding/json
// escapes all five by default, which is why this writer exists.
func encodeString(s string) string {
	var b strings.Builder
	b.WriteByte('"')
	for _, r := range utf16Units(s) {
		switch {
		case r == '"':
			b.WriteString(`\"`)
		case r == '\\':
			b.WriteString(`\\`)
		case r == '\n':
			b.WriteString(`\n`)
		case r == '\r':
			b.WriteString(`\r`)
		case r == '\t':
			b.WriteString(`\t`)
		case r == '\b':
			b.WriteString(`\b`)
		case r == '\f':
			b.WriteString(`\f`)
		case r < 0x20:
			fmt.Fprintf(&b, `\u%04x`, r)
		case r >= 0xd800 && r <= 0xdfff:
			// Lone surrogate (paired ones were recombined by utf16Units).
			fmt.Fprintf(&b, `\u%04x`, r)
		default:
			b.WriteRune(rune(r))
		}
	}
	b.WriteByte('"')
	return b.String()
}

// utf16Units yields code points, recombining valid surrogate pairs, so lone
// surrogates can be spotted and escaped exactly like JSON.stringify does.
func utf16Units(s string) []rune {
	units := utf16.Encode([]rune(s))
	out := make([]rune, 0, len(units))
	for i := 0; i < len(units); i++ {
		u := units[i]
		if u >= 0xd800 && u <= 0xdbff && i+1 < len(units) && units[i+1] >= 0xdc00 && units[i+1] <= 0xdfff {
			out = append(out, utf16.DecodeRune(rune(u), rune(units[i+1])))
			i++
			continue
		}
		out = append(out, rune(u))
	}
	// Guard against invalid UTF-8 in the source producing replacement chars.
	if !utf8.ValidString(s) {
		return []rune(s)
	}
	return out
}

// canonicalBody renders with object keys SORTED, arrays untouched, numbers verbatim.
func canonicalBody(v any, indent int) string {
	pad := strings.Repeat("  ", indent)
	padIn := strings.Repeat("  ", indent+1)
	switch t := v.(type) {
	case nil:
		return "null"
	case rawNumber:
		return string(t)
	case bool:
		if t {
			return "true"
		}
		return "false"
	case string:
		return encodeString(t)
	case []any:
		if len(t) == 0 {
			return "[]"
		}
		parts := make([]string, len(t))
		for i, e := range t {
			parts[i] = padIn + canonicalBody(e, indent+1)
		}
		return "[\n" + strings.Join(parts, ",\n") + "\n" + pad + "]"
	case *orderedObject:
		if len(t.keys) == 0 {
			return "{}"
		}
		keys := append([]string(nil), t.keys...)
		sort.Strings(keys)
		parts := make([]string, len(keys))
		for i, k := range keys {
			parts[i] = padIn + encodeString(k) + ": " + canonicalBody(t.vals[k], indent+1)
		}
		return "{\n" + strings.Join(parts, ",\n") + "\n" + pad + "}"
	default:
		panic(fmt.Sprintf("canonicalBody: unexpected %T", v))
	}
}

// orderReport emits one line per object, "<path>\t<keys in original order>".
// Traversal is by SORTED keys so both sides walk identical paths even when
// their orders differ.
func orderReport(v any, path string, lines *[]string) {
	switch t := v.(type) {
	case []any:
		for i, e := range t {
			orderReport(e, fmt.Sprintf("%s[%d]", path, i), lines)
		}
	case *orderedObject:
		*lines = append(*lines, path+"\t"+strings.Join(t.keys, ","))
		keys := append([]string(nil), t.keys...)
		sort.Strings(keys)
		for _, k := range keys {
			orderReport(t.vals[k], path+"."+k, lines)
		}
	}
}

func canonicalize(wire []byte) (body string, order string, err error) {
	parsed, err := parsePreservingNumbers(wire)
	if err != nil {
		return "", "", err
	}
	var lines []string
	orderReport(parsed, "$", &lines)
	return canonicalBody(parsed, 0) + "\n", strings.Join(lines, "\n") + "\n", nil
}

// marshalWire serializes a payload the way it goes on the wire, with Go's
// HTML escaping disabled so the bytes are comparable to JSON.stringify.
func marshalWire(payload any) ([]byte, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(payload); err != nil {
		return nil, err
	}
	return bytes.TrimRight(buf.Bytes(), "\n"), nil
}

File: difftest/pi/canon.mjs
// Shared canonicalizer for the pi-vs-Go differential.
//
// Both sides feed their *serialized wire JSON* through an equivalent
// canonicalizer (this one; go/canon.go is the line-for-line twin) and the
// runner diffs the results. See ../README.md for exactly what is normalized
// and what is deliberately left alone.

const RAW = Symbol("raw-number");

/**
 * Parse JSON while preserving each number's original literal text, so that
 * `1024` vs `1024.0` vs `1e3` stays visible instead of being laundered
 * through a float round-trip.
 */
export function parsePreservingNumbers(text) {
	return JSON.parse(text, function (key, value, context) {
		if (typeof value === "number" && context && typeof context.source === "string") {
			return { [RAW]: context.source };
		}
		if (typeof value === "number") {
			// Fallback for runtimes without source access.
			return { [RAW]: String(value) };
		}
		return value;
	});
}

function isRawNumber(v) {
	return v !== null && typeof v === "object" && Object.getOwnPropertySymbols(v).includes(RAW);
}

// JSON string escaping matched to JSON.stringify (ES2019 well-formed):
// escape the mandatory control set and lone surrogates, and nothing else.
// Notably U+2028/U+2029 are NOT escaped, and <, >, & are NOT escaped.
export function encodeString(s) {
	let out = '"';
	for (let i = 0; i < s.length; i++) {
		const c = s.charCodeAt(i);
		const ch = s[i];
		if (ch === '"') out += '\\"';
		else if (ch === "\\") out += "\\\\";
		else if (ch === "\n") out += "\\n";
		else if (ch === "\r") out += "\\r";
		else if (ch === "\t") out += "\\t";
		else if (ch === "\b") out += "\\b";
		else if (ch === "\f") out += "\\f";
		else if (c < 0x20) out += "\\u" + c.toString(16).padStart(4, "0");
		else if (c >= 0xd800 && c <= 0xdbff) {
			const next = s.charCodeAt(i + 1);
			if (next >= 0xdc00 && next <= 0xdfff) {
				out += ch + s[i + 1];
				i++;
			} else out += "\\u" + c.toString(16).padStart(4, "0"); // lone high surrogate
		} else if (c >= 0xdc00 && c <= 0xdfff) {
			out += "\\u" + c.toString(16).padStart(4, "0"); // lone low surrogate
		} else out += ch;
	}
	return out + '"';
}

/** Canonical body: object keys sorted, arrays untouched, numbers verbatim. */
export function canonicalBody(v, indent = 0) {
	const pad = "  ".repeat(indent);
	const padIn = "  ".repeat(indent + 1);
	if (v === null) return "null";
	if (isRawNumber(v)) return v[RAW];
	if (typeof v === "boolean") return String(v);
	if (typeof v === "string") return encodeString(v);
	if (Array.isArray(v)) {
		if (v.length === 0) return "[]";
		return "[\n" + v.map((e) => padIn + canonicalBody(e, indent + 1)).join(",\n") + "\n" + pad + "]";
	}
	const keys = Object.keys(v).sort();
	if (keys.length === 0) return "{}";
	return (
		"{\n" +
		keys.map((k) => padIn + encodeString(k) + ": " + canonicalBody(v[k], indent + 1)).join(",\n") +
		"\n" +
		pad +
		"}"
	);
}

/**
 * Key-order report: one line per object, `<path>\t<keys in original order>`.
 * Traversal itself is by SORTED keys so both sides walk identical paths even
 * when their orders differ — a pure order difference shows up in the value
 * column, never as a phantom structural difference.
 */
export function orderReport(v, path = "$", lines = []) {
	if (v === null || isRawNumber(v) || typeof v !== "object") return lines;
	if (Array.isArray(v)) {
		v.forEach((e, i) => orderReport(e, `${path}[${i}]`, lines));
		return lines;
	}
	const original = Object.keys(v);
	lines.push(`${path}\t${original.join(",")}`);
	for (const k of [...original].sort()) orderReport(v[k], `${path}.${k}`, lines);
	return lines;
}

export function canonicalize(wireText) {
	const parsed = parsePreservingNumbers(wireText);
	return {
		body: canonicalBody(parsed) + "\n",
		order: orderReport(parsed).join("\n") + "\n",
	};
}
